x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java

```diff
@@ -573,21 +616,24 @@ void merge(MergeableIndicesPrivilege other) {
             }
         }
 
-        private static void collatePrivilegesByIndices(
-            IndicesPrivileges[] indicesPrivileges,
-            boolean allowsRestrictedIndices,
-            Map<Set<String>, MergeableIndicesPrivilege> indicesPrivilegesMap
+        private static void collatePrivilegesByIndicesAndClusterAliases(
+            final IndicesPrivileges[] indicesPrivileges,
+            final boolean allowsRestrictedIndices,
+            final Map<Set<String>, Map<Set<String>, MergeableIndicesPrivilege>> indicesPrivilegesMapByClusterAlias
         ) {
             for (final IndicesPrivileges indicesPrivilege : indicesPrivileges) {
-                // if a index privilege is an explicit denial, then we treat it as non-existent since we skipped these in the past when
+                // if an index privilege is an explicit denial, then we treat it as non-existent since we skipped these in the past when
                 // merging
                 final boolean isExplicitDenial = indicesPrivileges.length == 1
                     && "none".equalsIgnoreCase(indicesPrivilege.getPrivileges()[0]);
                 if (isExplicitDenial || (indicesPrivilege.allowRestrictedIndices() != allowsRestrictedIndices)) {
                     continue;
                 }
-                final Set<String> key = newHashSet(indicesPrivilege.getIndices());
-                indicesPrivilegesMap.compute(key, (k, value) -> {
+                final var clusterAlias = indicesPrivilege.hasRemoteClusters()
+                    ? newHashSet(indicesPrivilege.getRemoteClusters())
+                    : LOCAL_CLUSTER_ALIAS_KEY;
+                final Set<String> indicesKey = newHashSet(indicesPrivilege.getIndices());
+                indicesPrivilegesMapByClusterAlias.computeIfAbsent(clusterAlias, k -> new HashMap<>()).compute(indicesKey, (k, value) -> {
```

Comment:
The `computeIfAbsent` pattern is pretty repetitive: we have maps representing rather similar things at three different layers: 

- `CompositeRolesStore`: `Map<Set<String>, Map<Set<String>, MergeableIndicesPrivilege>>`
- `RemoteIndicesPermission.Builder`: `Map<Set<String>, List<IndicesPermission.Group>>`
- `Role`: `Map<Set<String>, List<IndicesPermissionGroupDefinition>>`

At each layer, we essentially just translate the values of the map. I don't like the redundancy of adding each entry one by one to a new map at each layer but haven't come up with a cleaner way to do this.